IDelegateCreationExpression

333fred opened the pull request:
Implements IDelegateCreationExpression, which in some cases replaces IConversionExpression. Fixes https://github.com/dotnet/roslyn/issues/20095.

333fred commented:
Note that a couple of parent pointer tests will fail in VB. @heejaechang's work in https://github.com/dotnet/roslyn/pull/21857 will fix this.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,19 @@
+    public interface IDelegateCreationExpression : IOperation
+    {
+        /// <summary>
+        /// The conversion and lambda body or method binding that this delegate is created from.
```
>The conversion and lambda body [](start = 12, length = 30)

Is this accurate? Also, could this ever be an invalid expression?

333fred replied:
I updated the comment, as it wasn't accurate on the conversion scenario anymore. Looking at my tests so far, it does not appear that this will ever be an IInvalidExpression. The only NoneOperations we have the moment here are BoundQueryLambdas, which https://github.com/dotnet/roslyn/pull/21356 will address by turning them into IAnonymousFunctionExpressions.

333fred replied:
As a follow up, it appears that for truly invalid syntax, such as `new Action(gibberish)` or `new Action(Method1, Method2)`, this fails to bind entirely, and we return an `IInvalidExpression`. I'll add some tests for this.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -697,13 +700,26 @@ private IOperation CreateBoundConversionOperation(BoundConversion boundConversio
+                    // return an IDelegateCreationExpression instead of a conversion.
+                    if (boundConversion.Type.InheritsSpecialType(SpecialType.System_Delegate))
+                    {
+                        return new LazyDelegateCreationExpression(operand, _semanticModel, syntax, type, constantValue, isImplicit);
```
>return new LazyDelegateCreationExpression(operand, _semanticModel, syntax, type, constantValue, isImplicit); [](start = 24, length = 108)

Do we want to get here for error cases? Do we have tests hitting this code path?

AlekseyTs replied:
Do we have tests hitting this code path for error cases?

---
In reply to: [138489201](https://github.com/dotnet/roslyn/pull/22056#discussion_r138489201) [](ancestors = 138489201)

333fred replied:
We do. All of the invalid Lambda tests hit this path.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -697,13 +700,26 @@ private IOperation CreateBoundConversionOperation(BoundConversion boundConversio
+                {
+                    // We need to determine if this is a conversion to a delegate type. If it is, it's a delegate creation, so we
+                    // return an IDelegateCreationExpression instead of a conversion.
+                    if (boundConversion.Type.InheritsSpecialType(SpecialType.System_Delegate))
```
>.InheritsSpecialType(SpecialType.System_Delegate) [](start = 44, length = 49)

There is already a helper called IsDelegateType, compiler uses it to detect delegate types.

333fred replied:
Ah, missed it because it's on TypeSymbol, I was only looking at ITypeSymbol. I'll remove this.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -722,6 +738,41 @@ private IConversionExpression CreateBoundAsOperatorOperation(BoundAsOperator bou
+            Lazy<IOperation> target = new Lazy<IOperation>(() =>
+            {
+                if (boundDelegateCreationExpression.Argument is BoundMethodGroup boundMethodGroup &&
+                    boundMethodGroup.Methods.Length == 1)
```
>boundMethodGroup.Methods.Length == 1 [](start = 20, length = 36)

I do not understand why do we need to have this restriction? What if there are several candidates?

333fred replied:
We have this restriction because we don't have support for unresolved BoundMethodGroups in V1. There's nowhere for me to expose the other methods that were considered.

AlekseyTs replied:
> We have this restriction because we don't have support for unresolved BoundMethodGroups in V1. There's nowhere for me to expose the other methods that were considered.

I am not sure I am following, couldn't there be a success case when there is more than one candidate. What happens then? Do we have a test for a scenario like that?

333fred replied:
I've updated this condition to check to see if `BoundDelegateCreationExpression.MethodOpt` is non-null. We're still not exposing other candidates if they were considered, but we'll only expose an actual method binding if a single method was selected at compile time. This won't cover a few valid scenarios (such as dynamic), but we need an actual `IMethodGroup` or something similar to solve that.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -722,6 +738,41 @@ private IConversionExpression CreateBoundAsOperatorOperation(BoundAsOperator bou
+            bool isVirtual = methodSymbol != null && (methodSymbol.IsAbstract || methodSymbol.IsOverride || methodSymbol.IsVirtual) && !suppressVirtualCalls;
+            Lazy<IOperation> instance = new Lazy<IOperation>(() => Create(boundMethodGroup.InstanceOpt));
+            SyntaxNode bindingSyntax = boundMethodGroup.Syntax;
+            ITypeSymbol bindingType = boundMethodGroup.Type;
```
>ITypeSymbol bindingType = boundMethodGroup.Type; [](start = 12, length = 48)

I think ```bindingType``` should always be null.

333fred replied:
Fixed.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -722,6 +738,41 @@ private IConversionExpression CreateBoundAsOperatorOperation(BoundAsOperator bou
+            ITypeSymbol bindingType = boundMethodGroup.Type;
+            Optional<object> bindingConstantValue = ConvertToOptional(boundMethodGroup.ConstantValue);
+            bool isImplicit = boundMethodGroup.WasCompilerGenerated;
+            return new LazyMethodBindingExpression(methodSymbol, isVirtual, instance, methodSymbol, _semanticModel, bindingSyntax, bindingType, bindingConstantValue, boundMethodGroup.WasCompilerGenerated);
```
>methodSymbol [](start = 51, length = 12)

Why are we passing methodSymbol twice? Could it be null, what happens then?

333fred replied:
We're passing it twice because IMethodBindingExpression has a MethodSymbol property, with Type IMethodSymbol, and it inherits from IMemberReferenceExpression, which has a Member property with Type ISymbol. I'll deduplicate it.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -545,27 +547,72 @@ Namespace Microsoft.CodeAnalysis.Semantics
+                boundConversion.Operand.Kind = BoundKind.RelaxationLambda OrElse
+                boundConversion.Operand.Kind = BoundKind.UnboundLambda OrElse
+                boundConversion.Operand.Kind = BoundKind.DelegateCreationExpression) AndAlso
+               type.InheritsSpecialType(SpecialType.System_Delegate) Then
```
>InheritsSpecialType [](start = 20, length = 19)

I believe we already have a helper IsDelegateType

333fred replied:
Fixed.

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions.cs:
```diff
@@ -100,5 +93,24 @@ internal static INamespaceSymbol GetNestedNamespace(this INamespaceSymbol contai
+        /// <summary>
+        /// Returns whether this type inherits from the given WellKnownType
+        /// </summary>
+        internal static bool InheritsSpecialType(this ITypeSymbol originalType, SpecialType specialType)
```
>InheritsSpecialType [](start = 29, length = 19)

It doesn't look like there is a real need in this method.

333fred replied:
Removed.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -545,27 +547,72 @@ Namespace Microsoft.CodeAnalysis.Semantics
+            ElseIf boundConversion.ConversionKind = VisualBasic.ConversionKind.Identity AndAlso
+                   boundConversion.Operand.Kind = BoundKind.DelegateCreationExpression Then
+                ' In this scenario, we're something like New DelegateType(AddressOf Method)
+                ' We don't want to process both this conversion and the DelegateCreationExpression
```
>We don't want to process both this conversion and the DelegateCreationExpression [](start = 18, length = 80)

This is probably true only if the conversion is implicit?

AlekseyTs replied:
Could you point me to the test that is affected by this logic? 

---
In reply to: [138492959](https://github.com/dotnet/roslyn/pull/22056#discussion_r138492959) [](ancestors = 138492959)

AlekseyTs replied:
I believe here and below we should have more targeted logic, let's discuss offline.

---
In reply to: [138494022](https://github.com/dotnet/roslyn/pull/22056#discussion_r138494022) [](ancestors = 138494022,138492959)

333fred replied:
I'll be in the office later today, I'll stop by then. As an example though, https://github.com/dotnet/roslyn/pull/22056/files#diff-076069470e21e72cef89138ff31fa9f4R658. We create a BoundConversion on top of a BoundDelegateCreation in the case of `New DelegateType(AddressOf Method)`. From my reading of the spec, this should be exactly equivalent to `AddressOf Method`, so we shouldn't have nested `IDelegateCreationExpression`s, like we do for `new Action((Action)Method)` in C#.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -545,27 +547,72 @@ Namespace Microsoft.CodeAnalysis.Semantics
-            Dim conversion = New Conversion(New KeyValuePair(Of ConversionKind, MethodSymbol)(boundConversion.ConversionKind, methodSymbol))
+            If (boundConversion.Operand.Kind = BoundKind.Lambda OrElse
+                boundConversion.Operand.Kind = BoundKind.QueryLambda OrElse
+                boundConversion.Operand.Kind = BoundKind.RelaxationLambda OrElse
```
>boundConversion.Operand.Kind = BoundKind.RelaxationLambda [](start = 16, length = 57)

Can an operand ever be a BoundRelaxationLambda? I believe it is always stored in BoundConversion.ExtendedInfoOpt and should always be ignored by IOperation.

333fred replied:
Fixed.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -545,27 +547,72 @@ Namespace Microsoft.CodeAnalysis.Semantics
+                boundConversion.Operand.Kind = BoundKind.QueryLambda OrElse
+                boundConversion.Operand.Kind = BoundKind.RelaxationLambda OrElse
+                boundConversion.Operand.Kind = BoundKind.UnboundLambda OrElse
+                boundConversion.Operand.Kind = BoundKind.DelegateCreationExpression) AndAlso
```
>boundConversion.Operand.Kind = BoundKind.DelegateCreationExpression [](start = 16, length = 67)

I'd like to better understand the scenarios when we are dropping conversions on top of delegate creation.

AlekseyTs replied:
Since this  condition is targeting the same case as above, let's make this condition more targeted as well.

---
In reply to: [138495420](https://github.com/dotnet/roslyn/pull/22056#discussion_r138495420) [](ancestors = 138495420)

333fred replied:
Updated.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -545,27 +547,72 @@ Namespace Microsoft.CodeAnalysis.Semantics
+            Dim type As ITypeSymbol = boundDelegateCreationExpression.Type
+            Dim constantValue As [Optional](Of Object) = ConvertToOptional(boundDelegateCreationExpression.ConstantValueOpt)
+            Dim isImplicit As Boolean = boundDelegateCreationExpression.WasCompilerGenerated
+            Return New LazyMethodBindingExpression(method, isVirtual, instance, method, _semanticModel, syntax, type, constantValue, isImplicit)
```
>Return New LazyMethodBindingExpression(method, isVirtual, instance, method, _semanticModel, syntax, type, constantValue, isImplicit) [](start = 12, length = 132)

Same questions as for C#.

333fred replied:
Same answer as C#. I've deduplicated this.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IAnonymousFunctionExpression.cs:
```diff
@@ -33,9 +33,8 @@ static void F()
-    Initializer: IConversionExpression (Implicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: System.Action) (Syntax: '() => F()')
-        Conversion: CommonConversion (Exists: True, IsIdentity: False, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
-        Operand: IAnonymousFunctionExpression (Symbol: lambda expression) (OperationKind.AnonymousFunctionExpression, Type: null) (Syntax: '() => F()')
+    Initializer: IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.Action) (Syntax: '() => F()')
```
>IDelegateCreationExpression [](start = 17, length = 27)

Should this node be Implicit?

333fred replied:
Added a test for this.

---
In reply to: [138497185](https://github.com/dotnet/roslyn/pull/22056#discussion_r138497185) [](ancestors = 138497185)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,976 @@
+{
+    void Main()
+    {
+        Action a = /*<bind>*/new Action(() => { })/*</bind>*/;
```
>new Action( [](start = 29, length = 11)

We should have a test for a scenario when more than one argument is given to the delegate's constructor.

333fred replied:
Added.

AlekseyTs commented:
>         Action x /*<bind>*/= () => F()/*</bind>*/;

It would be good to test scenario with an explicit conversion

---
Refers to: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IAnonymousFunctionExpression.cs:24 in eedb899. [](commit_id = eedb89935c2aa78dd69fdaeb42f826d6685b4a8e, deletion_comment = False)

AlekseyTs commented:
>         Action<int> x /*<bind>*/= () => F()/*</bind>*/;

It would be good to test scenario with an explicit conversion

---
Refers to: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IAnonymousFunctionExpression.cs:101 in eedb899. [](commit_id = eedb89935c2aa78dd69fdaeb42f826d6685b4a8e, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IEventAssignmentExpression.cs:
```diff
@@ -38,8 +38,9 @@ void M()
       Instance Receiver: ILocalReferenceExpression: t (OperationKind.LocalReferenceExpression, Type: Test) (Syntax: 't')
-  Handler: IMethodBindingExpression: void C.Handler(System.Object sender, System.EventArgs e) (OperationKind.MethodBindingExpression, Type: System.EventHandler) (Syntax: 'Handler')
-      Instance Receiver: IInstanceReferenceExpression (OperationKind.InstanceReferenceExpression, Type: C) (Syntax: 'Handler')
+  Handler: IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.EventHandler) (Syntax: 'Handler')
```
>IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.EventHandler) (Syntax: 'Handler') [](start = 11, length = 117)

Should this nod be Implicit?

333fred replied:
It is. Added a test with just `Handler` to verify.

---
In reply to: [138681742](https://github.com/dotnet/roslyn/pull/22056#discussion_r138681742) [](ancestors = 138681742)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,976 @@
+IVariableDeclarationStatement (1 declarations) (OperationKind.VariableDeclarationStatement) (Syntax: 'Action a = () => { };')
+  IVariableDeclaration (1 variables) (OperationKind.VariableDeclaration) (Syntax: 'Action a = () => { };')
+    Variables: Local_1: System.Action a
+    Initializer: IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.Action) (Syntax: '() => { }')
```
>Initializer: IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.Action) (Syntax: '() => { }') [](start = 4, length = 126)

Should this node be Implicit.

333fred replied:
It is. Added a separate test that attempts to GetOperation just for the lambda expression to verify that we're only seeing the IAnonymousFunctionExpression node, not the wrapping IDelegateCreationExpression.

---
In reply to: [138728156](https://github.com/dotnet/roslyn/pull/22056#discussion_r138728156) [](ancestors = 138728156)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,976 @@
+
+            // Currently, GetOperation returns a straight BoundMethodGroup for the IdentifierNameSyntax. We can't directly
+            // convert a BoundMethodGroup to an IDelegateCreationExpression, so this returns OperationKind.None. When Heejae's
+            // rewrite of GetOperation is complete, this should return the IDelegateCreationExpression.
```
>this should return the IDelegateCreationExpression. [](start = 52, length = 51)

Please open an issue to come back to this test and update the baseline, link the unit-test to the issue.

333fred replied:
Heejae's stuff is pushed, just removing the comment.

---
In reply to: [138730510](https://github.com/dotnet/roslyn/pull/22056#discussion_r138730510) [](ancestors = 138730510)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,976 @@
+{
+    void Main()
+    {
+        Action a = /*<bind>*/M1/*</bind>*/;
```
>Action a = /*<bind>*/M1/*</bind>*/; [](start = 8, length = 35)

Please add a separate unit-test that verifies tree for the entire statement if using just an expression node doesn't work, otherwise the scenario isn't tested at all.

333fred replied:
Done.

---
In reply to: [138731012](https://github.com/dotnet/roslyn/pull/22056#discussion_r138731012) [](ancestors = 138731012)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,976 @@
+
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void DelegateCreationExpression_ImplicitMethodBinding_InvalidIdentifier()
```
>DelegateCreationExpression_ImplicitMethodBinding_InvalidIdentifier [](start = 20, length = 66)

Same comments as for DelegateCreationExpression_ImplicitMethodBinding

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,976 @@
+
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void DelegateCreationExpression_ImplicitMethodBinding_InvalidReturnType()
```
>DelegateCreationExpression_ImplicitMethodBinding_InvalidReturnType [](start = 20, length = 66)

Same comments as for DelegateCreationExpression_ImplicitMethodBinding

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,976 @@
+
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void DelegateCreationExpression_ImplicitMethodBinding_InvalidArgumentType()
```
>DelegateCreationExpression_ImplicitMethodBinding_InvalidArgumentType [](start = 20, length = 68)

Same comments as for DelegateCreationExpression_ImplicitMethodBinding

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,976 @@
+{
+    void Main()
+    {
+        Action a = /*<bind>*/(Action)M1/*</bind>*/;
```
>(Action)M1 [](start = 29, length = 10)

For this, and similar error cases below, it would be good to add a receiver (not ```this```), so that we could observe what happens with it.

333fred replied:
Added. We were hiding the receivers because we didn't have an implementation of `IBoundNodeWithChildren.Children` for `BoundMethodGroup`. I added an implementation which exposes the reference to the receiver.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,976 @@
+{
+    void Main()
+    {
+        Action a = /*<bind>*/(Action)M1/*</bind>*/;
```
>M1 [](start = 37, length = 2)

It would be good to add a test when there is more than one candidate and none of them is a good match.

333fred replied:
Added. Also added multiple candidates and one match.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,976 @@
+{
+    void Main()
+    {
+        Action a = /*<bind>*/new Action(M1)/*</bind>*/;
```
>M1 [](start = 40, length = 2)

Would we get any information about receiver, when there is one?

333fred replied:
Added a test with a non-implicit receiver.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,976 @@
+            VerifyOperationTreeAndDiagnosticsForTest<ObjectCreationExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        // TODO: Invalid tests for below function
```
>TODO [](start = 11, length = 4)

Please create an issue to track this work and remove the TODO.

333fred replied:
This was just a reminder I left myself after leaving for the day, forgot to remove it.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,976 @@
+{
+    void Main()
+    {
+        Action a = /*<bind>*/new Action((Action)M1)/*</bind>*/;
```
>(Action)M1 [](start = 40, length = 10)

It would be good to test scenarios when the explicit cast is fine, but the new fails due to shape mismatch.

333fred replied:
Added.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/Binding/BindingObjectInitializerTests.vb:
```diff
@@ -1119,9 +1119,8 @@ IObjectCreationExpression (Constructor: Sub C1..ctor()) (OperationKind.ObjectCre
-            Right: IConversionExpression (Implicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: System.Func(Of System.Int32)) (Syntax: 'Function()  ... nd Function')
-                Conversion: CommonConversion (Exists: True, IsIdentity: False, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
-                Operand: IAnonymousFunctionExpression (Symbol: Function () As System.Int32) (OperationKind.AnonymousFunctionExpression, Type: null) (Syntax: 'Function()  ... nd Function')
+            Right: IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.Func(Of System.Int32)) (Syntax: 'Function()  ... nd Function')
```
>Right: IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type:  [](start = 12, length = 84)

Should this node be Implicit?

333fred replied:
I would think so. IsImplicit for VB conversions wasn't taking into account whether the cast was explicit in code. I've updated this code, and added a version of this test that binds just the Function() body to make sure it returns just the function body.

---
In reply to: [138739482](https://github.com/dotnet/roslyn/pull/22056#discussion_r138739482) [](ancestors = 138739482)

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IConversionExpression.vb:
```diff
@@ -758,9 +758,8 @@ End Module]]>.Value
-    Initializer: IConversionExpression (Implicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: System.Action(Of System.Int32)) (Syntax: 'Sub(i As In ... End Sub')
-        Conversion: CommonConversion (Exists: True, IsIdentity: False, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
-        Operand: IAnonymousFunctionExpression (Symbol: Sub (i As System.Int32)) (OperationKind.AnonymousFunctionExpression, Type: null) (Syntax: 'Sub(i As In ... End Sub')
+    Initializer: IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.Action(Of System.Int32)) (Syntax: 'Sub(i As In ... End Sub')
```
>Initializer: IDelegateCreationExpression (OperationKind.DelegateCreationExpression, [](start = 4, length = 83)

Should this node be Implicit.

333fred replied:
Added a test for this.

---
In reply to: [138740561](https://github.com/dotnet/roslyn/pull/22056#discussion_r138740561) [](ancestors = 138740561)

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IConversionExpression.vb:
```diff
@@ -1061,7 +1058,9 @@ End Module]]>.Value
   IVariableDeclaration (1 variables) (OperationKind.VariableDeclaration) (Syntax: 'a')
     Variables: Local_1: a As System.Action
-    Initializer: IOperation:  (OperationKind.None) (Syntax: 'AddressOf M2')
+    Initializer: IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.Action) (Syntax: 'AddressOf M2')
```
>Initializer: IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.Action) (Syntax: 'AddressOf M2') [](start = 4, length = 129)

Should this node be implicit?

333fred replied:
It should be. Added a test to verify just the initializer section.

---
In reply to: [138741293](https://github.com/dotnet/roslyn/pull/22056#discussion_r138741293) [](ancestors = 138741293)

AlekseyTs commented:
>         Public Sub ConversionExpression_Implicit_WideningMethodGroupToDelegate_RelaxationArguments_InvalidConversion()

Please add a similar test with a receiver, we should verify how is it getting exposed.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IConversionExpression.vb:1010 in eedb899. [](commit_id = eedb89935c2aa78dd69fdaeb42f826d6685b4a8e, deletion_comment = False)

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IConversionExpression.vb:
```diff
@@ -2348,24 +2349,81 @@ IVariableDeclarationStatement (1 declarations) (OperationKind.VariableDeclaratio
         End Sub
 
+        <Fact()>
+        Public Sub ConversionExpression_Implicit_WideningDelegateTypeConversion()
```
>ConversionExpression_Implicit_WideningDelegateTypeConversion [](start = 19, length = 60)

It would be good to cover an Anonymous Delegate conversions as well.

333fred replied:
Added.

---
In reply to: [138743990](https://github.com/dotnet/roslyn/pull/22056#discussion_r138743990) [](ancestors = 138743990)

AlekseyTs commented:
>     Partial Public Class IOperationTests

Please make sure we have coverage for explicit CType/DirectCast/TryCast conversions to delegate types.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IConversionExpression.vb:9 in eedb899. [](commit_id = eedb89935c2aa78dd69fdaeb42f826d6685b4a8e, deletion_comment = False)

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:
```diff
@@ -0,0 +1,966 @@
+End Module]]>.Value
+
+            Dim expectedOperationTree = <![CDATA[
+IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.Action) (Syntax: 'Sub() Conso ... iteLine("")')
```
Should this node be Implicit?

333fred replied:
Added a test for this scenario.

---
In reply to: [138745351](https://github.com/dotnet/roslyn/pull/22056#discussion_r138745351) [](ancestors = 138745351)

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:
```diff
@@ -0,0 +1,966 @@
+        ILabeledStatement (Label: exit) (OperationKind.LabeledStatement) (Syntax: 'Function() 1')
+          Statement: null
+        IReturnStatement (OperationKind.ReturnStatement) (Syntax: 'Function() 1')
+          ReturnedValue: ILocalReferenceExpression:  (OperationKind.LocalReferenceExpression, Type: System.Object) (Syntax: 'Function() 1')
```
>ReturnedValue: ILocalReferenceExpression:  (OperationKind.LocalReferenceExpression, Type: System.Object) (Syntax: 'Function() 1') [](start = 10, length = 129)

It is not clear what local does this node reference.

333fred replied:
This is apparently a SynthesizedLocal (of type FunctionReturnValue) introduced by the LambdaBodyBinder.

---
In reply to: [138746756](https://github.com/dotnet/roslyn/pull/22056#discussion_r138746756) [](ancestors = 138746756)

AlekseyTs replied:
Let's open an issue to investigate whether we can/should remove these compiler generated artifacts like this extra return, the exit label, etc 

---
In reply to: [141468218](https://github.com/dotnet/roslyn/pull/22056#discussion_r141468218) [](ancestors = 141468218,138746756)

333fred replied:
This is probably an extension on [https://github.com/dotnet/roslyn/issues/7397](https://github.com/dotnet/roslyn/issues/7397).

---
In reply to: [141977566](https://github.com/dotnet/roslyn/pull/22056#discussion_r141977566) [](ancestors = 141977566,141468218,138746756)

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:
```diff
@@ -0,0 +1,966 @@
+  Target: IAnonymousFunctionExpression (Symbol: Function () As System.Object) (OperationKind.AnonymousFunctionExpression, Type: null) (Syntax: 'Function() 1')
+      IBlockStatement (3 statements, 1 locals) (OperationKind.BlockStatement) (Syntax: 'Function() 1')
+        Locals: Local_1: <anonymous local> As System.Object
+        IReturnStatement (OperationKind.ReturnStatement) (Syntax: '1')
```
>IReturnStatement (OperationKind.ReturnStatement) (Syntax: '1') [](start = 8, length = 62)

Should this node be Implicit?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:
```diff
@@ -0,0 +1,966 @@
+          ReturnedValue: IConversionExpression (Implicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: System.Object) (Syntax: '1')
+              Conversion: CommonConversion (Exists: True, IsIdentity: False, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+              Operand: ILiteralExpression (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1) (Syntax: '1')
+        ILabeledStatement (Label: exit) (OperationKind.LabeledStatement) (Syntax: 'Function() 1')
```
>ILabeledStatement (Label: exit) (OperationKind.LabeledStatement) (Syntax: 'Function() 1') [](start = 8, length = 89)

Should this node be Implicit?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:
```diff
@@ -0,0 +1,966 @@
+              Operand: ILiteralExpression (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1) (Syntax: '1')
+        ILabeledStatement (Label: exit) (OperationKind.LabeledStatement) (Syntax: 'Function() 1')
+          Statement: null
+        IReturnStatement (OperationKind.ReturnStatement) (Syntax: 'Function() 1')
```
>IReturnStatement (OperationKind.ReturnStatement) (Syntax: 'Function() 1') [](start = 8, length = 73)

Should this node be Implicit?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:
```diff
@@ -0,0 +1,966 @@
+        ILabeledStatement (Label: exit) (OperationKind.LabeledStatement) (Syntax: 'Function() 1')
+          Statement: null
+        IReturnStatement (OperationKind.ReturnStatement) (Syntax: 'Function() 1')
+          ReturnedValue: ILocalReferenceExpression:  (OperationKind.LocalReferenceExpression, Type: System.Object) (Syntax: 'Function() 1')
```
>ReturnedValue: ILocalReferenceExpression:  (OperationKind.LocalReferenceExpression, Type: System.Object) (Syntax: 'Function() 1') [](start = 10, length = 129)

Should this node be Implicit?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:
```diff
@@ -0,0 +1,966 @@
+Namespace Microsoft.CodeAnalysis.VisualBasic.UnitTests.Semantics
+
+    Partial Public Class IOperationTests
+        Inherits SemanticModelTestBase
```
>SemanticModelTestBase [](start = 17, length = 21)

Scenarios that are using CType should also be tested with DirectCast/TryCast

333fred replied:
Added.

---
In reply to: [138749030](https://github.com/dotnet/roslyn/pull/22056#discussion_r138749030) [](ancestors = 138749030)

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:
```diff
@@ -0,0 +1,966 @@
+End Module]]>.Value
+
+            Dim expectedOperationTree = <![CDATA[
+IConversionExpression (Implicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: System.Func(Of System.String), IsInvalid) (Syntax: 'New Func(Of ... Of Method2)')
```
It feels like there is an inconsistency with scenario when lambdas are involved, those are using IDelegateCreation even for error cases. Might be worth investigating why.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -545,27 +547,72 @@ Namespace Microsoft.CodeAnalysis.Semantics
+            ElseIf boundConversion.ConversionKind = VisualBasic.ConversionKind.Identity AndAlso
+                   boundConversion.Operand.Kind = BoundKind.DelegateCreationExpression Then
+                ' In this scenario, we're something like New DelegateType(AddressOf Method)
+                ' We don't want to process both this conversion and the DelegateCreationExpression
```
As we discussed, let's check that conversion.SyntaxNode is object creation expression and that delegateCreation.SyntaxNode is AddressOf expression

---
In reply to: [138711440](https://github.com/dotnet/roslyn/pull/22056#discussion_r138711440) [](ancestors = 138711440)

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:
```diff
@@ -0,0 +1,966 @@
+Namespace Microsoft.CodeAnalysis.VisualBasic.UnitTests.Semantics
+
+    Partial Public Class IOperationTests
+        Inherits SemanticModelTestBase
```
>SemanticModelTestBase [](start = 17, length = 21)

Also New DelegateType with two or more arguments.

333fred replied:
Added.

---
In reply to: [138755797](https://github.com/dotnet/roslyn/pull/22056#discussion_r138755797) [](ancestors = 138755797)

AlekseyTs commented:
Done with review pass (iteration 16)

333fred commented:
> It would be good to test scenario with an explicit conversion
I can't reply this these comments directly, but I have tests for explicit conversions on top of lambda expressions in IOperationTests_IDelegateCreationExpression.cs.

333fred commented:
@AlekseyTs I've addressed your C# related comments. I'll work on VB on the plane ride tomorrow and push the changes when I get to the east coast.

333fred commented:
> Please add a similar test with a receiver, we should verify how is it getting exposed.
> Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IConversionExpression.vb:9 in eedb899. [](commit_id = eedb899, deletion_comment = False)

Done. This creates a BoundTypeOrValueExpression, which we currently don't handle and expose a NoneOperation.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/Statement.cs:
```diff
@@ -21,14 +21,35 @@ internal partial class BoundBadStatement
     }
 
-    partial class BoundFixedStatement
+    partial class BoundMethodGroup
```
>BoundMethodGroup [](start = 18, length = 16)

This node is an expression, it belongs in Expressions.cs. Also, I believe [https://github.com/dotnet/roslyn/pull/21857](https://github.com/dotnet/roslyn/pull/21857) already includes proper implementation.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -722,6 +735,42 @@ private IConversionExpression CreateBoundAsOperatorOperation(BoundAsOperator bou
+
+        private IMethodBindingExpression CreateBoundMethodGroupSingleMethodOperation(BoundMethodGroup boundMethodGroup, IMethodSymbol methodSymbol, bool suppressVirtualCalls)
+        {
+            bool isVirtual = methodSymbol != null && (methodSymbol.IsAbstract || methodSymbol.IsOverride || methodSymbol.IsVirtual) && !suppressVirtualCalls;
```
>methodSymbol != null [](start = 29, length = 20)

Do we ever expect methodSymbol to be null here?

333fred replied:
I don't believe we do. Fixing.

---
In reply to: [140097698](https://github.com/dotnet/roslyn/pull/22056#discussion_r140097698) [](ancestors = 140097698)

AlekseyTs commented on src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs:
```diff
@@ -2989,16 +3037,16 @@ internal abstract partial class MemberReferenceExpression : Operation, IMemberRe
-        public BaseMethodBindingExpression(IMethodSymbol method, bool isVirtual, ISymbol member, SemanticModel semanticModel, SyntaxNode syntax, ITypeSymbol type, Optional<object> constantValue, bool isImplicit) :
-            base(member, OperationKind.MethodBindingExpression, semanticModel, syntax, type, constantValue, isImplicit)
+        public BaseMethodBindingExpression(IMethodSymbol method, bool isVirtual, SemanticModel semanticModel, SyntaxNode syntax, ITypeSymbol type, Optional<object> constantValue, bool isImplicit) :
+            base(null, OperationKind.MethodBindingExpression, semanticModel, syntax, type, constantValue, isImplicit)
```
>null [](start = 17, length = 4)

Passing null here feels wrong. It looks like that will result in IMemberReferenceExpression.Symbol to be null. I think we should pass ```method``` instead.

AlekseyTs replied:
> It looks like that will result in IMemberReferenceExpression.Member to be null. 

I see that ```MemberReferenceExpression.Member``` is overridden below to prevent that. Still, I think it is better to pass ```method``` and, instead of making the ```Member``` property virtual, implement ```Method``` like ```public IMethodSymbol Method => (IMethodSymbol)Member;```. This way we won't have a need for a separate field to store the method.

---
In reply to: [140099464](https://github.com/dotnet/roslyn/pull/22056#discussion_r140099464) [](ancestors = 140099464)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,1418 @@
+
+            // Currently, GetOperation returns a straight BoundMethodGroup for the IdentifierNameSyntax. We can't directly
+            // convert a BoundMethodGroup to an IDelegateCreationExpression, so this returns OperationKind.None. When Heejae's
+            // rewrite of GetOperation is complete, this should return the IDelegateCreationExpression.
```
>rewrite of GetOperation is complete, this should return the IDelegateCreationExpression. [](start = 15, length = 88)

This is probably incorrect because IDelegateCreationExpression is supposed to be marked as Implicit and getting IOperation for a syntax node isn't supposed to return implicit nodes.

333fred replied:
This is correct. I have separate tests to verify just the method binding, as well as the whole tree here.

---
In reply to: [140102041](https://github.com/dotnet/roslyn/pull/22056#discussion_r140102041) [](ancestors = 140102041)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,1418 @@
+}
+";
+            string expectedOperationTree = @"
+IInvalidExpression (OperationKind.InvalidExpression, Type: System.Action, IsInvalid) (Syntax: 'new Action(M1)')
```
It feels like there is still some degree of inconsistency. I would expect to get tree similar to the one in DelegateCreationExpression_ExplicitMethodBinding_InvalidReturnType. That might require making changes to the binder to make sure it doesn't create general bad expression node.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,976 @@
+{
+    void Main()
+    {
+        Action a = /*<bind>*/new Action((Action)M1)/*</bind>*/;
```
>>It would be good to test scenarios when the explicit cast is fine, but the new fails due to shape mismatch.

>Added.

I don't see a test like that. I see error scenarios that fail because implicit conversion from result of ```new``` expression to the target type doesn't exist, but not the scenario where binding of  the ```new``` expression itself fails, even though the explicit cast, used as its argument, binds successfully. 



---
In reply to: [139040080](https://github.com/dotnet/roslyn/pull/22056#discussion_r139040080) [](ancestors = 139040080)

333fred replied:
Ah, I misunderstood which part of the shape you were referring to. Adding a new test for this.

---
In reply to: [140106607](https://github.com/dotnet/roslyn/pull/22056#discussion_r140106607) [](ancestors = 140106607,139040080)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,1418 @@
+";
+            string expectedOperationTree = @"
+IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.Action<System.String>) (Syntax: 'new Action<string>(M1)')
+  Target: IOperation:  (OperationKind.None) (Syntax: 'M1')
```
> Target: IOperation:  (OperationKind.None) (Syntax: 'M1') [](start = 1, length = 57)

This is unexpected. It looks like there are no errors, why are we not getting the method reference?

333fred replied:
I'm not sure what you mean? We are getting a method reference.

---
In reply to: [140107408](https://github.com/dotnet/roslyn/pull/22056#discussion_r140107408) [](ancestors = 140107408)

AlekseyTs replied:
> I'm not sure what you mean? We are getting a method reference.

We didn't in iteration 22.

---
In reply to: [141477343](https://github.com/dotnet/roslyn/pull/22056#discussion_r141477343) [](ancestors = 141477343,140107408)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,1418 @@
+}
+";
+            string expectedOperationTree = @"
+IInvalidExpression (OperationKind.InvalidExpression, Type: System.Action<System.Int32>, IsInvalid) (Syntax: 'new Action<int>(M1)')
```
The result looks inconsistent with DelegateCreationExpression_ImplicitMethodBinding_MultipleCandidates_InvalidNoMatch test.

333fred replied:
Similar to your comment above, we'll have to make changes to the binder to fix this.

---
In reply to: [140107720](https://github.com/dotnet/roslyn/pull/22056#discussion_r140107720) [](ancestors = 140107720)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,1418 @@
+  IVariableDeclaration (1 variables) (OperationKind.VariableDeclaration, IsInvalid) (Syntax: 'Action<int> a = M1;')
+    Variables: Local_1: System.Action<System.Int32> a
+    Initializer: IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.Action<System.Int32>, IsInvalid) (Syntax: 'M1')
+        Target: IMethodBindingExpression: void Program.M1(System.Object o) (OperationKind.MethodBindingExpression, Type: null, IsInvalid) (Syntax: 'M1')
```
>Target: IMethodBindingExpression: void Program.M1(System.Object o) (OperationKind.MethodBindingExpression, Type: null, IsInvalid) (Syntax: 'M1') [](start = 8, length = 144)

This feels incorrect, the method isn't a match.

333fred replied:
The issue here was that the BoundMethodGroup itself is fine, but the conversion sitting on top of it isn't. To resolve, I'm filtering the method group when created. If the conversion sitting on top `HasErrors`, then I fall back to the CSharpOperationFactory, which returns a NoneOperation here.

---
In reply to: [140108124](https://github.com/dotnet/roslyn/pull/22056#discussion_r140108124) [](ancestors = 140108124)

AlekseyTs commented:
Done with review pass (iteration 22).

333fred commented:
>     Partial Public Class IOperationTests

We do, in IOperationTests_IDelegateCreationExpression.vb

---
In reply to: [329302183](https://github.com/dotnet/roslyn/pull/22056#issuecomment-329302183) [](ancestors = 329302183)

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IConversionExpression.vb:9 in eedb899. [](commit_id = eedb89935c2aa78dd69fdaeb42f826d6685b4a8e, deletion_comment = False)

333fred commented:
>         Action x /*<bind>*/= () => F()/*</bind>*/;

Added.

---
In reply to: [329234005](https://github.com/dotnet/roslyn/pull/22056#issuecomment-329234005) [](ancestors = 329234005)

---
Refers to: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IAnonymousFunctionExpression.cs:24 in eedb899. [](commit_id = eedb89935c2aa78dd69fdaeb42f826d6685b4a8e, deletion_comment = False)

333fred commented:
>         Action<int> x /*<bind>*/= () => F()/*</bind>*/;

Added, both for a scenario with a valid conversion and an invalid conversion.

---
In reply to: [329234091](https://github.com/dotnet/roslyn/pull/22056#issuecomment-329234091) [](ancestors = 329234091)

---
Refers to: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IAnonymousFunctionExpression.cs:101 in eedb899. [](commit_id = eedb89935c2aa78dd69fdaeb42f826d6685b4a8e, deletion_comment = False)

333fred commented:
@AlekseyTs I think this is ready for another review.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,1591 @@
+
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void DelegateCreationExpression_ImplicitMethodBinding_InvalidIdentifer()
```
>Identifer [](start = 76, length = 9)

Typo "Identifer"?

333fred replied:
Fixed.

---
In reply to: [141978739](https://github.com/dotnet/roslyn/pull/22056#discussion_r141978739) [](ancestors = 141978739)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,1591 @@
+";
+            string expectedOperationTree = @"
+IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.Action, IsInvalid) (Syntax: '(Action)M1')
+  Target: IOperation:  (OperationKind.None, IsInvalid) (Syntax: 'M1')
```
It feels like the baseline change since iteration 22 is undesired. The tree became inconsistent with the one in DelegateCreationExpression_ImplicitMethodBinding_InvalidReturnType test.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,1591 @@
+";
+            string expectedOperationTree = @"
+IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.Action, IsInvalid) (Syntax: '(Action)p.M1')
+  Target: IOperation:  (OperationKind.None, IsInvalid) (Syntax: 'p.M1')
```
>Target: IOperation:  (OperationKind.None, IsInvalid) (Syntax: 'p.M1') [](start = 2, length = 69)

Same comment as for the previous test.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,1591 @@
+";
+            string expectedOperationTree = @"
+IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.Action, IsInvalid) (Syntax: 'new Action(M1)')
+  Target: IOperation:  (OperationKind.None, IsInvalid) (Syntax: 'M1')
```
>Target: IOperation:  (OperationKind.None, IsInvalid) (Syntax: 'M1') [](start = 2, length = 67)

I would expect to get a tree similar to DelegateCreationExpression_ImplicitMethodBinding_InvalidReturnType

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,1591 @@
+IInvalidExpression (OperationKind.InvalidExpression, Type: System.Action, IsInvalid) (Syntax: 'new Action((Action)M1)')
+  Children(1):
+      IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.Action, IsInvalid) (Syntax: '(Action)M1')
+        Target: IOperation:  (OperationKind.None, IsInvalid) (Syntax: 'M1')
```
>Target: IOperation:  (OperationKind.None, IsInvalid) (Syntax: 'M1') [](start = 8, length = 67)

Losing the method reference feels unexpected.

AlekseyTs commented:
>         Dim a As Func(Of String) = DirectCast(Function() 1, Func(Of String))'BIND:"DirectCast(Function() 1, Func(Of String))"

It would be useful to test scenario when conversion for the return type doesn't exist (vs. disallowed with Option Strict On).

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:600 in 84ff26f. [](commit_id = 84ff26fa750c5503921d84537cb30123204a029e, deletion_comment = False)

AlekseyTs commented:
>         Dim a As Action(Of Object) = New Action(Of Object)(Sub(i As Integer) Console.WriteLine())'BIND:"New Action(Of Object)(Sub(i As Integer) Console.WriteLine())"

It would be good to test scenario when conversion doesn't exist rather than disallowed.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:1095 in 84ff26f. [](commit_id = 84ff26fa750c5503921d84537cb30123204a029e, deletion_comment = False)

AlekseyTs commented:
>         Public Sub DelegateCreationExpression_DirectCastAddressOf_InvalidArgumentType()

It doesn't look like there is an invalid argument type in this scenario. Consider adjusting the name of the unit-test. I think several unit-tests have this problem.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:1689 in 84ff26f. [](commit_id = 84ff26fa750c5503921d84537cb30123204a029e, deletion_comment = False)

AlekseyTs commented:
> IConversionExpression (Explicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: System.Action) (Syntax: 'New Action( ... Of Method2)')

The change in baseline between iterations 22 and 45 feels unexpected. 

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:2119 in 84ff26f. [](commit_id = 84ff26fa750c5503921d84537cb30123204a029e, deletion_comment = False)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +591,99 @@ Namespace Microsoft.CodeAnalysis.Semantics
+            Return validConversionKind AndAlso validConversionSyntax AndAlso validOperandSyntax
+        End Function
+
+        Private Function CreateConversionOperation(operand As BoundNode, conversionKind As ConversionKind, conversionSyntax As SyntaxNode) As (methodSymbol As MethodSymbol, operation As Lazy(Of IOperation), isAddressOfDelegateCreation As Boolean)
```
>Private Function CreateConversionOperation(operand As BoundNode, conversionKind As ConversionKind, conversionSyntax As SyntaxNode) As (methodSymbol As MethodSymbol, operation As Lazy(Of IOperation), isAddressOfDelegateCreation As Boolean) [](start = 8, length = 238)

This line is too long. It is also hard to read because it elements of the returned tuple look like parameters of the method. Consider putting the return type on a separate line. Something like:
```
Private Function CreateConversionOperation( ...
) As (methodSymbol As MethodSymbol, operation As Lazy(Of IOperation), isAddressOfDelegateCreation As Boolean)
...
```

333fred replied:
Done.

---
In reply to: [142262835](https://github.com/dotnet/roslyn/pull/22056#discussion_r142262835) [](ancestors = 142262835)

AlekseyTs commented:
> IConversionExpression (Explicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: System.Action) (Syntax: 'New Action( ... Of Method2)')

I am not sure why do we have IConversionExpression here.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:2050 in 84ff26f. [](commit_id = 84ff26fa750c5503921d84537cb30123204a029e, deletion_comment = False)

333fred commented:
>         Public Sub DelegateCreationExpression_DirectCastAddressOf_InvalidArgumentType()

Ok. I'll rename these tests and add tests that actually cover this scenario.

---
In reply to: [333666812](https://github.com/dotnet/roslyn/pull/22056#issuecomment-333666812) [](ancestors = 333666812)

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:1689 in 84ff26f. [](commit_id = 84ff26fa750c5503921d84537cb30123204a029e, deletion_comment = False)

333fred commented:
> IConversionExpression (Explicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: System.Action) (Syntax: 'New Action( ... Of Method2)')

Fat-fingered a refactor. Fixed.

---
In reply to: [333673895](https://github.com/dotnet/roslyn/pull/22056#issuecomment-333673895) [](ancestors = 333673895)

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:2050 in 84ff26f. [](commit_id = 84ff26fa750c5503921d84537cb30123204a029e, deletion_comment = False)

333fred commented:
> IConversionExpression (Explicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: System.Action) (Syntax: 'New Action( ... Of Method2)')

Same as above.

---
In reply to: [333668271](https://github.com/dotnet/roslyn/pull/22056#issuecomment-333668271) [](ancestors = 333668271)

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:2119 in 84ff26f. [](commit_id = 84ff26fa750c5503921d84537cb30123204a029e, deletion_comment = False)

333fred commented:
>         Dim a As Action(Of Object) = New Action(Of Object)(Sub(i As Integer) Console.WriteLine())'BIND:"New Action(Of Object)(Sub(i As Integer) Console.WriteLine())"

Added a test for this, renamed existing test to be more appropriate.

---
In reply to: [333264557](https://github.com/dotnet/roslyn/pull/22056#issuecomment-333264557) [](ancestors = 333264557)

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:1095 in 84ff26f. [](commit_id = 84ff26fa750c5503921d84537cb30123204a029e, deletion_comment = False)

333fred commented:
>         Dim a As Func(Of String) = DirectCast(Function() 1, Func(Of String))'BIND:"DirectCast(Function() 1, Func(Of String))"

Added this test.

---
In reply to: [333263614](https://github.com/dotnet/roslyn/pull/22056#issuecomment-333263614) [](ancestors = 333263614)

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:600 in 84ff26f. [](commit_id = 84ff26fa750c5503921d84537cb30123204a029e, deletion_comment = False)

333fred commented:
@AlekseyTs addressed feedback.

AlekseyTs commented:
>     Initializer: IConversionExpression (Implicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: System.Action(Of System.String), IsInvalid) (Syntax: 'AddressOf Method2')

This feels somewhat inconsistent with the tree that we are getting for an explicit conversion case (DelegateCreationExpression_CTypeAddressOf_InvalidArgumentType))
```
Dim a As Action(Of String) = CType(AddressOf Method2, Action(Of String))'BIND:"CType(AddressOf Method2, Action(Of String))"
```
```
            Dim expectedOperationTree = <![CDATA[
IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.Action(Of System.String), IsInvalid) (Syntax: 'CType(Addre ... Of String))')
  Target: IOperation:  (OperationKind.None, IsInvalid) (Syntax: 'AddressOf Method2')
      Children(1):
          IOperation:  (OperationKind.None, IsInvalid) (Syntax: 'Method2')
            Children(1):
                IInstanceReferenceExpression (OperationKind.InstanceReferenceExpression, Type: M1, IsInvalid) (Syntax: 'Method2')
]]>.Value
```

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:1708 in f38e07b. [](commit_id = f38e07b1b3447f0ad940e635a6a769ef58d1cac1, deletion_comment = False)

AlekseyTs commented:
>     Initializer: IConversionExpression (Implicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: System.Action, IsInvalid) (Syntax: 'AddressOf Method2')

Should this be a IDelegateCreationExpression instead?

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:1669 in f38e07b. [](commit_id = f38e07b1b3447f0ad940e635a6a769ef58d1cac1, deletion_comment = False)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_Delegates.vb:
```diff
@@ -134,6 +134,16 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                                boundFirstArgument = boundConversion.Update(boundConversion.Operand,
+                                                                            boundConversion.ConversionKind,
+                                                                            boundConversion.Checked,
+                                                                            True,
```
>True, [](start = 76, length = 5)

Consider adding a comment at the end of this line with the name of the parameter, or using named argument.

333fred replied:
Comment added.

---
In reply to: [142805013](https://github.com/dotnet/roslyn/pull/22056#discussion_r142805013) [](ancestors = 142805013)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +591,104 @@ Namespace Microsoft.CodeAnalysis.Semantics
+               ((boundConversion.Operand.Kind = BoundKind.Lambda OrElse
+                 boundConversion.Operand.Kind = BoundKind.QueryLambda OrElse
+                 boundConversion.Operand.Kind = BoundKind.UnboundLambda) AndAlso
+                boundConversion.Type.IsDelegateType()) Then
```
It looks like this condition is used in several places, consider extracting it into a helper method. Alternatively, the Type can be passed into CreateConversionOperation, which can perform all these checks and calculate isDelegateCreation rather than just calculating isAddressOfDelegateCreation.

333fred replied:
Refactored.

---
In reply to: [142806631](https://github.com/dotnet/roslyn/pull/22056#discussion_r142806631) [](ancestors = 142806631)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +591,104 @@ Namespace Microsoft.CodeAnalysis.Semantics
             Return New LazyVisualBasicConversionExpression(operand, conversion, isExplicit, isTryCast, isChecked, _semanticModel, syntax, type, constantValue, isImplicit)
         End Function
 
+        Private Function IsDelegateCreationSyntax(conversionKind As ConversionKind, conversionSyntax As SyntaxNode, operandSyntax As SyntaxNode) As Boolean
```
>Private Function [](start = 8, length = 16)

Can it be Shared (i.e. not an instance) function?

333fred replied:
Yes, done.

---
In reply to: [142807928](https://github.com/dotnet/roslyn/pull/22056#discussion_r142807928) [](ancestors = 142807928)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +591,104 @@ Namespace Microsoft.CodeAnalysis.Semantics
+            Dim syntax As SyntaxNode = boundDelegateCreationExpression.Syntax
+            Dim type As ITypeSymbol = boundDelegateCreationExpression.Type
+            Dim constantValue As [Optional](Of Object) = ConvertToOptional(boundDelegateCreationExpression.ConstantValueOpt)
+            Dim isImplicit As Boolean = boundDelegateCreationExpression.WasCompilerGenerated OrElse boundDelegateCreationExpression.Syntax.Kind() = SyntaxKind.AddressOfExpression
```
>OrElse boundDelegateCreationExpression.Syntax.Kind() = SyntaxKind.AddressOfExpression [](start = 93, length = 85)

Given that tests do not exposed value for IsImplicit, it is not clear what is the impact of this condition. It would also be good to add a clarifying comment. We should really consider an alternative to modify Binder to mark the node as compiler generated instead.

AlekseyTs replied:
Ok, given how the CreateBoundDelegateCreationExpressionChildOperation is implemented, it feels like we should always mark this node as Implicit. The reason is - the operand is going to be using the same syntax node and only one node should be considered explicit for the given syntax node.

---
In reply to: [142809527](https://github.com/dotnet/roslyn/pull/22056#discussion_r142809527) [](ancestors = 142809527)

333fred replied:
Done, and added a comment.

---
In reply to: [142820740](https://github.com/dotnet/roslyn/pull/22056#discussion_r142820740) [](ancestors = 142820740,142809527)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +591,104 @@ Namespace Microsoft.CodeAnalysis.Semantics
+            Dim instance As Lazy(Of IOperation) =
+                        New Lazy(Of IOperation)(Function() Create(If(boundDelegateCreationExpression.ReceiverOpt, boundDelegateCreationExpression.MethodGroupOpt?.ReceiverOpt)))
+            Dim syntax As SyntaxNode = boundDelegateCreationExpression.Syntax
+            Dim type As ITypeSymbol = boundDelegateCreationExpression.Type
```
>Dim type As ITypeSymbol = boundDelegateCreationExpression.Type [](start = 12, length = 62)

I think method reference should have type null.

333fred replied:
Done.

---
In reply to: [142810033](https://github.com/dotnet/roslyn/pull/22056#discussion_r142810033) [](ancestors = 142810033)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +591,104 @@ Namespace Microsoft.CodeAnalysis.Semantics
+                                               Not boundDelegateCreationExpression.SuppressVirtualCalls
+            Dim instance As Lazy(Of IOperation) =
+                        New Lazy(Of IOperation)(Function() Create(If(boundDelegateCreationExpression.ReceiverOpt, boundDelegateCreationExpression.MethodGroupOpt?.ReceiverOpt)))
+            Dim syntax As SyntaxNode = boundDelegateCreationExpression.Syntax
```
>Dim syntax As SyntaxNode = boundDelegateCreationExpression.Syntax [](start = 12, length = 65)

It is not clear why this is the right syntax to use for the method reference. For example, why not use syntax of the MethodGroupOpt, when on is available?

AlekseyTs replied:
This is probably something about strange bound tree shape around AddressOf. Comments would be very helpful if this is the case.

---
In reply to: [142810723](https://github.com/dotnet/roslyn/pull/22056#discussion_r142810723) [](ancestors = 142810723)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +591,104 @@ Namespace Microsoft.CodeAnalysis.Semantics
+            Dim syntax As SyntaxNode = boundDelegateCreationExpression.Syntax
+            Dim type As ITypeSymbol = boundDelegateCreationExpression.Type
+            Dim constantValue As [Optional](Of Object) = ConvertToOptional(boundDelegateCreationExpression.ConstantValueOpt)
+            Dim isImplicit As Boolean = boundDelegateCreationExpression.WasCompilerGenerated
```
>boundDelegateCreationExpression.WasCompilerGenerated [](start = 40, length = 52)

Similar to the syntax, should we use value from the MethodGroupOpt?

AlekseyTs commented:
>         Dim a As Action = New Action(Sub() Console.WriteLine())'BIND:"New Action(Sub() Console.WriteLine())"

Do we have a test for similar scenario with AddressOf (i.e. an exact signature match)?

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:1279 in f38e07b. [](commit_id = f38e07b1b3447f0ad940e635a6a769ef58d1cac1, deletion_comment = False)

AlekseyTs commented:
>         Dim a As Action = New Action(Sub() Console.WriteLine())'BIND:"New Action(Sub() Console.WriteLine())"

Found it - DelegateCreationExpression_DelegateCreationAddressOfArgument

---
In reply to: [334311851](https://github.com/dotnet/roslyn/pull/22056#issuecomment-334311851) [](ancestors = 334311851)

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:1279 in f38e07b. [](commit_id = f38e07b1b3447f0ad940e635a6a769ef58d1cac1, deletion_comment = False)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +591,104 @@ Namespace Microsoft.CodeAnalysis.Semantics
+                                               Not boundDelegateCreationExpression.SuppressVirtualCalls
+            Dim instance As Lazy(Of IOperation) =
+                        New Lazy(Of IOperation)(Function() Create(If(boundDelegateCreationExpression.ReceiverOpt, boundDelegateCreationExpression.MethodGroupOpt?.ReceiverOpt)))
+            Dim syntax As SyntaxNode = boundDelegateCreationExpression.Syntax
```
>Dim syntax As SyntaxNode = boundDelegateCreationExpression.Syntax [](start = 12, length = 65)

Consider adding a comment that binder creates BoundDelegateCreationExpression node for AddressOf expression and that is the node we want use for the operand of IDelegateCreationExpression.

333fred replied:
Done.

---
In reply to: [142816374](https://github.com/dotnet/roslyn/pull/22056#discussion_r142816374) [](ancestors = 142816374)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +591,104 @@ Namespace Microsoft.CodeAnalysis.Semantics
+            End If
+        End Function
+
+        Private Function CreateBoundDelegateCreationExpressionOperation(boundDelegateCreationExpression As BoundDelegateCreationExpression) As IDelegateCreationExpression
```
>CreateBoundDelegateCreationExpressionOperation [](start = 25, length = 46)

What test hits this method?

333fred replied:
One example is DelegateCreationExpression_ImplicitAddressOf.

---
In reply to: [142816525](https://github.com/dotnet/roslyn/pull/22056#discussion_r142816525) [](ancestors = 142816525)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +591,104 @@ Namespace Microsoft.CodeAnalysis.Semantics
+            If (conversionKind And VisualBasic.ConversionKind.UserDefined) = VisualBasic.ConversionKind.UserDefined Then
+                Dim userDefinedConversion As BoundUserDefinedConversion = DirectCast(operand, BoundUserDefinedConversion)
+                Return (userDefinedConversion.Call.Method, New Lazy(Of IOperation)(Function() Create(userDefinedConversion.Operand)), isAddressOfDelegateCreation:=False)
+            ElseIf IsDelegateCreationSyntax(conversionKind, conversionSyntax, operand.Syntax) Then
```
>IsDelegateCreationSyntax(conversionKind, conversionSyntax, operand.Syntax)  [](start = 19, length = 75)

I do not think this syntax check is sufficient. We should at least check if the target type is a delegate type. Please add a test where an AddressOf is converted to something other than a delegate type.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +591,104 @@ Namespace Microsoft.CodeAnalysis.Semantics
+        Private Function IsDelegateCreationSyntax(conversionKind As ConversionKind, conversionSyntax As SyntaxNode, operandSyntax As SyntaxNode) As Boolean
+            ' An identity conversion is introduced by the compiler on top of New DelegateType(AddressOf Method). This node
+            ' only exists for the convenience of SemanticModel, so we want to classify it as a Delegate Creation
+            Dim validConversionKind = conversionKind = ConversionKind.Identity OrElse
```
>validConversionKind [](start = 16, length = 19)

I am not sure what difference does conversion kind make. If 
- the target type is a delegate type 
- the operand's syntax is an AddressOfExpression 
- the operands bound node is one of the: BoundKind.DelegateCreationExpression, BoundKind.MethodGroup, BoundKind.AddressOfOperator, or BoundKind.BadExpression (I guess we should check that its child is AddressOfOperator with the same syntax)

then this is a delegate creation.

333fred replied:
Fixed up.

---
In reply to: [142823813](https://github.com/dotnet/roslyn/pull/22056#discussion_r142823813) [](ancestors = 142823813)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +591,104 @@ Namespace Microsoft.CodeAnalysis.Semantics
+                Else
+                    ' This is an error scenario in which we have a delegate conversion, but it failed to bind correctly, so the child here
+                    ' is either a bad expression or a method group. Delegate to standard operation handling for this.
+                    Debug.Assert(operand.Kind = BoundKind.MethodGroup OrElse
```
>operand.Kind = BoundKind.MethodGroup [](start = 33, length = 36)

Could you point me a to a test case, for which this condition is true?

333fred replied:
I believe this was true earlier, but in later revisions this part of the condition is never hit.

As part of the CreateConversionOperand refactor I just did, this branch is no longer just for exceptional code, and can be hit by lambdas as well. Because of this, I've removed the assert altogether.

---
In reply to: [142995679](https://github.com/dotnet/roslyn/pull/22056#discussion_r142995679) [](ancestors = 142995679)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +591,104 @@ Namespace Microsoft.CodeAnalysis.Semantics
+                                      (conversionKind And (Not ConversionKind.DelegateRelaxationLevelMask)) = 0
+
+            ' Any of the explicit cast types, as well as New DelegateType(AddressOf Method)
+            Dim validConversionSyntax = conversionSyntax.Kind() = SyntaxKind.CTypeExpression OrElse
```
>validConversionSyntax [](start = 16, length = 21)

I guess we are missing the case of an implicit conversion, i.e. when the conversion node and the operand are both associated with the same AddressOfExpressionSyntax

333fred replied:
Added.

---
In reply to: [142996089](https://github.com/dotnet/roslyn/pull/22056#discussion_r142996089) [](ancestors = 142996089)

333fred commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +591,104 @@ Namespace Microsoft.CodeAnalysis.Semantics
+                    ' This is an error scenario in which we have a delegate conversion, but it failed to bind correctly, so the child here
+                    ' is either a bad expression or a method group. Delegate to standard operation handling for this.
+                    Debug.Assert(operand.Kind = BoundKind.MethodGroup OrElse
+                                 operand.Kind = BoundKind.BadExpression OrElse
```
As above, the assert has been removed altogether. However, ConversionExpression_Implicit_InvalidMethodGroupToDelegateSyntax would have hit it.

---
In reply to: [142995710](https://github.com/dotnet/roslyn/pull/22056#discussion_r142995710) [](ancestors = 142995710)

TyOverby commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests.vb:
```diff
@@ -1,4 +1,4 @@
-' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+﻿' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
Please investigate the BOM issues

333fred replied:
This is correct, as it's adding the BOM.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +583,105 @@ Namespace Microsoft.CodeAnalysis.Semantics
+                                                  conversionSyntax.Kind() = SyntaxKind.TryCastExpression OrElse
+                                                  conversionSyntax.Kind() = SyntaxKind.ObjectCreationExpression OrElse
+                                                  (conversionSyntax.Kind() = SyntaxKind.AddressOfExpression AndAlso
+                                                   ReferenceEquals(conversionSyntax, operand.Syntax))) AndAlso
```
>ReferenceEquals(conversionSyntax, operand.Syntax) [](start = 51, length = 49)

```conversionSyntax Is operand.Syntax``` ?

333fred replied:
Sorry, still not a native VB speaker.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +583,105 @@ Namespace Microsoft.CodeAnalysis.Semantics
+                                                   ReferenceEquals(conversionSyntax, operand.Syntax))) AndAlso
+                                                 operand.Syntax.Kind() = SyntaxKind.AddressOfExpression
+
+            Dim validLambdaConversionSyntax = operand.Kind = BoundKind.Lambda OrElse
```
>Syntax [](start = 37, length = 6)

It looks like condition isn't about syntax.

333fred replied:
It was, forgot to rename it when I refactored.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +583,105 @@ Namespace Microsoft.CodeAnalysis.Semantics
+                                                  conversionSyntax.Kind() = SyntaxKind.ObjectCreationExpression OrElse
+                                                  (conversionSyntax.Kind() = SyntaxKind.AddressOfExpression AndAlso
+                                                   ReferenceEquals(conversionSyntax, operand.Syntax))) AndAlso
+                                                 operand.Syntax.Kind() = SyntaxKind.AddressOfExpression
```
>operand.Syntax.Kind() = SyntaxKind.AddressOfExpression [](start = 49, length = 54)

I think it would be good to strengthen this condition with ```operand.Kind = BoundKind.DelegateCreationExpression```, or replace it with this condition.

333fred replied:
Unfortunately, the operand there isn't always a BoundDelegateCreationExpression. In my tests, it can also be a BoundAddressOfOperator or a BoundBadExpression, which is why I just went with a syntax check
I can swap that out for checking each of the 3 operand kinds, but it seemed simpler to me to just have a syntax check.

AlekseyTs commented:
>           IMethodReferenceExpression: Function System.Object.ToString() As System.String (IsVirtual) (OperationKind.MethodReferenceExpression, Type: null) (Syntax: 'AddressOf o.ToString')

I think it would be good to assert inside the test visitor that IMethodReferenceExpression.Type is always null.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:1858 in 24365db. [](commit_id = 24365db52512f6d41e11b2f7cdd50438af1a9f5c, deletion_comment = False)

333fred commented:
Assert added.

333fred commented:
@AlekseyTs feedback addressed.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ITupleExpression.vb:
```diff
@@ -531,19 +516,13 @@ Class C
-IConversionExpression (Implicit, TryCast: False, Unchecked) (OperatorMethod: Function C.op_Implicit(x As (System.Int32, System.String)) As C) (OperationKind.ConversionExpression, Type: C) (Syntax: '(0, Nothing)')
-  Conversion: CommonConversion (Exists: True, IsIdentity: False, IsNumeric: False, IsReference: False, IsUserDefined: True) (MethodSymbol: Function C.op_Implicit(x As (System.Int32, System.String)) As C)
-  Operand: 
-    IConversionExpression (Implicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: (System.Int32, System.Object)) (Syntax: '(0, Nothing)')
```
>IConversionExpression (Implicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: (System.Int32, System.Object)) (Syntax: '(0, Nothing)') [](start = 4, length = 158)

It feels like there is something wrong with this conversion node. It converts from (System.Int32, System.Object) to (System.Int32, System.Object), but says it is not an identity conversion. We probably shouldn't have this node in the tree at all. Please open an issue to follow up. We will have to create a new unit-test that gets the tree for the entire Dim statement.

333fred replied:
https://github.com/dotnet/roslyn/issues/22587

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/AnonymousTypesTests.vb:
```diff
@@ -586,13 +574,13 @@ End Module]]>.Value
 IAnonymousObjectCreationExpression (OperationKind.AnonymousObjectCreationExpression, Type: <anonymous type: a As System.Object>) (Syntax: 'New With {.a = Nothing}')
   Initializers(1):
-      ISimpleAssignmentExpression (OperationKind.SimpleAssignmentExpression, Type: System.Object, Constant: null) (Syntax: '.a = Nothing')
+      ISimpleAssignmentExpression (OperationKind.SimpleAssignmentExpression, Type: System.Object, Constant: null, IsImplicit) (Syntax: '.a = Nothing')
```
>IsImplicit [](start = 114, length = 10)

It is not clear why this node became Implicit, feels unexpected.

333fred replied:
Ok, so the reason is this line:
http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Operations/VisualBasicOperationFactory_Methods.vb,200

value in this example is the BoundConversion from Nothing to Object, which is implicit. Previously, this conversion was not implicit, so the node wasn't marked as implicit. I believe this is a bug in GetObjectCreationInitializers, and not related to this PR. Filed https://github.com/dotnet/roslyn/issues/22588.

cston commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,1684 @@
+
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void DelegateCreationExpression_ImplicitMethodBinding_InvalidReturnType_InitializerBindingReturnsJustNoneOperation()
```
It looks like binding returns an `IOperation` other than `None`.

333fred replied:
Yes, forgot to update the test name when we updated the output.

---
In reply to: [143323041](https://github.com/dotnet/roslyn/pull/22056#discussion_r143323041) [](ancestors = 143323041)

cston commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.cs:
```diff
@@ -0,0 +1,1684 @@
+  Target: 
+    IDelegateCreationExpression (OperationKind.DelegateCreationExpression, Type: System.Action) (Syntax: '(Action)(() => { })')
+      Target: 
+        IAnonymousFunctionExpression (Symbol: lambda expression) (OperationKind.AnonymousFunctionExpression, Type: null) (Syntax: '() => { }')
```
Are nested `IDelegateCreationExpression` objects expected?

cston commented on src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs:
```diff
@@ -3154,29 +3202,29 @@ internal abstract partial class MemberReferenceExpression : Operation, IMemberRe
         /// Referenced member.
         /// </summary>
-        public ISymbol Member { get; }
+        public virtual ISymbol Member { get; }
```
Is `Member` overridden?

333fred replied:
Not anymore, removed.

---
In reply to: [143323698](https://github.com/dotnet/roslyn/pull/22056#discussion_r143323698) [](ancestors = 143323698)

cston commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +583,108 @@ Namespace Microsoft.CodeAnalysis.Semantics
+        ''' is actually a delegate creation.
+        ''' </summary>
+        Private Function CreateConversionOperand(operand As BoundNode, conversionKind As ConversionKind, conversionSyntax As SyntaxNode, targetType As TypeSymbol
+                                                   ) As (methodSymbol As MethodSymbol, operation As Lazy(Of IOperation), isDelegateCreation As Boolean)
```
Consider using uppercase for first character of tuple element names.

333fred replied:
Done.

---
In reply to: [143323938](https://github.com/dotnet/roslyn/pull/22056#discussion_r143323938) [](ancestors = 143323938)

cston commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -517,29 +519,51 @@ Namespace Microsoft.CodeAnalysis.Semantics
+
+            Dim constantValue As [Optional](Of Object) = ConvertToOptional(boundTryCast.ConstantValueOpt)
+            Dim conversionInformation = CreateConversionOperand(boundTryCast.Operand, boundTryCast.ConversionKind, boundTryCast.Syntax, boundTryCast.Type)
+            Dim methodSymbol As MethodSymbol = conversionInformation.methodSymbol
```
`methodSymbol` is not used in this method or the next method.

333fred replied:
Fixed.

---
In reply to: [143330388](https://github.com/dotnet/roslyn/pull/22056#discussion_r143330388) [](ancestors = 143330388)

cston commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +583,108 @@ Namespace Microsoft.CodeAnalysis.Semantics
+                                              operand.Kind = BoundKind.QueryLambda OrElse
+                                              operand.Kind = BoundKind.UnboundLambda
+
+            Dim validTargetType = targetType.IsDelegateType()
```
Consider checking `targetType.IsDelegateType()` first and returning early if not `True`.

333fred replied:
Done

---
In reply to: [143330666](https://github.com/dotnet/roslyn/pull/22056#discussion_r143330666) [](ancestors = 143330666)

cston commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +583,108 @@ Namespace Microsoft.CodeAnalysis.Semantics
+                If operand.Kind = BoundKind.DelegateCreationExpression Then
+                    ' If the child is a BoundDelegateCreationExpression, we don't want to generate a nested IDelegateCreationExpression.
+                    ' So, the operand for the conversion will be the child of the BoundDelegateCreationExpression.
+                    operandLazy = New Lazy(Of IOperation)(Function() CreateBoundDelegateCreationExpressionChildOperation(DirectCast(operand, BoundDelegateCreationExpression)))
```
I don't follow this. Why do we need to avoid a nested `IDelegateCreationExpression` if the operand is also a `BoundDelegateCreationExpression`? From the tests, it looks like there are still various cases where we get nested instances.

333fred replied:
I'll add more to the comment, but the problem is that for `Dim x = New Action(AddressOf M2)`, this should be semantically equivalent to `Dim x = AddressOf M2`, but we generate both the BoundConversion and a BoundDelegateCreationExpression. So we need to compensate for that. I have tests to make sure that `Dim x = New Action(CType(AddressOf M2, Action))` is two nested IDelegateCreationExpressions, as we would expect.

---
In reply to: [143330842](https://github.com/dotnet/roslyn/pull/22056#discussion_r143330842) [](ancestors = 143330842)

cston commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -559,28 +583,108 @@ Namespace Microsoft.CodeAnalysis.Semantics
+                                               (method.IsAbstract OrElse method.IsOverride OrElse method.IsVirtual) AndAlso
+                                               Not boundDelegateCreationExpression.SuppressVirtualCalls
+            Dim instance As Lazy(Of IOperation) =
+                        New Lazy(Of IOperation)(Function() Create(If(boundDelegateCreationExpression.ReceiverOpt, boundDelegateCreationExpression.MethodGroupOpt?.ReceiverOpt)))
```
When is `boundDelegateCreationExpression.ReceiverOpt` null and `boundDelegateCreationExpression.MethodSymbol?.ReceiverOpt` not null?

333fred replied:
For this, see Binder_Delegates.vb:1049.

---
In reply to: [143331198](https://github.com/dotnet/roslyn/pull/22056#discussion_r143331198) [](ancestors = 143331198)

cston commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDelegateCreationExpression.vb:
```diff
@@ -0,0 +1,3404 @@
+
+Module Program
+    Sub Main()
+        Dim a As Action(Of Object) = DirectCast(Sub() Console.WriteLine(), Action(Of Object))'BIND:"DirectCast(Sub() 
```
Should the operand be a `IDelegateCreationExpression` so the `AddressOf` and the method symbol are included in the `IOperation`?

mavasani commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IAnonymousFunctionExpression.cs:
```diff
@@ -34,9 +34,88 @@ static void F()
+IAnonymousFunctionExpression (Symbol: lambda expression) (OperationKind.AnonymousFunctionExpression, Type: null) (Syntax: '() => F()')
+  IBlockStatement (2 statements) (OperationKind.BlockStatement, IsImplicit) (Syntax: 'F()')
+    IExpressionStatement (OperationKind.ExpressionStatement, IsImplicit) (Syntax: 'F()')
+      Expression: 
```
Unrelated to your PR, but probably we can avoid dumping the property name (and the associated newline) if an operation node has a single child which is always non-null:

```
IExpressionStatement (OperationKind.ExpressionStatement, IsImplicit) (Syntax: 'F()') 
  IInvocationExpression (void Program.F()) (OperationKind.InvocationExpression, Type: System.Void) (Syntax: 'F()') 
  Instance Receiver:
    null
  Arguments(0)
```

333fred commented:
Test failure is https://github.com/dotnet/roslyn/issues/21563

333fred commented:
@dotnet-bot retest windows_release_unit64_prtest please
